batched/system.go

```diff
@@ -0,0 +1,372 @@
+	"github.com/ipfs/go-datastore"
+	provider "github.com/ipfs/go-ipfs-provider"
+	"github.com/ipfs/go-ipfs-provider/queue"
+	"github.com/ipfs/go-ipfs-provider/simple"
+	logging "github.com/ipfs/go-log"
+	"github.com/ipfs/go-verifcid"
+	"github.com/multiformats/go-multihash"
+)
+
+var log = logging.Logger("provider.batched")
+
+type BatchProvidingSystem struct {
+	ctx     context.Context
+	close   context.CancelFunc
+	closewg sync.WaitGroup
+
+	reprovideInterval time.Duration
+	rsys              provideMany
+	keyProvider       simple.KeyChanFunc
+
+	q  *queue.Queue
+	ds datastore.Batching
+
+	dynamicCh chan cid.Cid
+
+	totalProvides, lastReprovideBatchSize     int
+	avgProvideDuration, lastReprovideDuration time.Duration
+}
+
+var _ provider.System = (*BatchProvidingSystem)(nil)
+
+type provideMany interface {
+	ProvideMany(ctx context.Context, keys []multihash.Multihash) error
+	Ready() bool
+}
+
+// Option defines the functional option type that can be used to configure
+// BatchProvidingSystem instances
+type Option func(system *BatchProvidingSystem) error
+
+var lastReprovideKey = datastore.NewKey("/provider/reprovide/lastreprovide")
+
+func New(provider provideMany, q *queue.Queue, opts ...Option) (*BatchProvidingSystem, error) {
+	s := &BatchProvidingSystem{
+		reprovideInterval: time.Hour * 24,
+		rsys:              provider,
+		keyProvider:       nil,
+		q:                 q,
+		ds:                datastore.NewMapDatastore(),
+		dynamicCh:         make(chan cid.Cid, 1),
+	}
+
+	for _, o := range opts {
+		if err := o(s); err != nil {
+			return nil, err
+		}
+	}
+
+	// This is after the options processing so we do not have to worry about leaking a context if there is an
+	// initialization error processing the options
+	ctx, cancel := context.WithCancel(context.Background())
+	s.ctx = ctx
+	s.close = cancel
+
+	return s, nil
+}
+
+func Datastore(batching datastore.Batching) Option {
+	return func(system *BatchProvidingSystem) error {
+		system.ds = batching
+		return nil
+	}
+}
+
+func ReproviderInterval(duration time.Duration) Option {
+	return func(system *BatchProvidingSystem) error {
+		system.reprovideInterval = duration
+		return nil
+	}
+}
+
+func KeyProvider(fn simple.KeyChanFunc) Option {
+	return func(system *BatchProvidingSystem) error {
+		system.keyProvider = fn
+		return nil
+	}
+}
+
+func (s *BatchProvidingSystem) Run() {
+	const pauseDetectionThreshold = time.Millisecond * 500
+	const maxCollectionDuration = time.Minute * 10
+	const maxTimeWaitingForProvides = time.Hour
+
+	provCh := s.q.Dequeue()
+
+	go func() {
+		s.closewg.Add(1)
+		defer s.closewg.Done()
+
+		m := make(map[cid.Cid]struct{})
+
+		maxCollectionDurationTimer := time.NewTimer(maxCollectionDuration)
+		pauseDetectTimer := time.NewTimer(maxTimeWaitingForProvides)
+		defer maxCollectionDurationTimer.Stop()
+		defer pauseDetectTimer.Stop()
+
+		for {
+			pauseDetectTimer.Reset(maxTimeWaitingForProvides)
+			maxCollectionDurationTimer.Reset(maxCollectionDuration)
```

Comment:
Thinking about this... we're going to loop through every 10 minutes, regardless of whether or not we actually have anything to do. The "correct" thing to do would be to only start time timer after we actually _have_ things to do (although that can get a bit tricky).

Reply:
That seems doable, it's not that different from what the `pauseDetectTimer` is doing.